Image Component

Blackbaud-BrandonJones opened the pull request:
Initial commit files and image component

Blackbaud-BobbyEarl commented on package.json:
```diff
@@ -0,0 +1,71 @@
+{
```
I'm seeing a lot of properties that probably don't need to be here.  https://github.com/blackbaud/skyux-indicators/blob/master/package.json as a comparison..

Blackbaud-BobbyEarl commented on CHANGELOG.md:
```diff
@@ -0,0 +1,3 @@
+# x.x.x (yyyy-mm-dd)
```
I realize you're working to release an RC.  Not sure if we should include this file yet or not.  Feels a little weird having it in its current state.

Blackbaud-BrandonJones replied:
Normally this file is included in the template.  Since things got a little out of whack when trying to get this repo set up, I am including the file in this PR as part of the "initial commit".

Blackbaud-BobbyEarl commented on src/app/app-extras.module.ts:
```diff
@@ -0,0 +1,24 @@
+
+@NgModule({
+  imports: [
+    SkyImageModule
```
I believe listing the `SkyImageModule` is sufficient.

Blackbaud-BrandonJones commented:
@Blackbaud-SteveBrush @Blackbaud-BobbyEarl I have updated the PR to include all unit and visual tests for the image component.  I fixed the issues with the package.json dependencies. This PR is ready for final review.

Blackbaud-SteveBrush commented on e2e/image.e2e.spec.ts:
```diff
@@ -0,0 +1,22 @@
+import {
```
Looks like this spec is a duplicate; probably can be removed.

Blackbaud-SteveBrush commented on package.json:
```diff
@@ -0,0 +1,33 @@
+  "license": "ISC",
+  "peerDependencies": {
+    "@angular/common": "^4.3.6",
+    "@angular/core": "^4.3.6",
```
All Angular peers should be `>=`.
```
"@angular/common": ">=4.3.6",
"@angular/core": ">=4.3.6"
```

Blackbaud-SteveBrush commented on skyuxconfig.json:
```diff
@@ -0,0 +1,27 @@
+{
+  "name": "skyux-media",
```
We should add the `"$schema"` property here, which will give the file type support in the IDE.
```
"$schema": "./node_modules/@skyux/config/skyuxconfig-schema.json",
```

Blackbaud-SteveBrush commented on src/app/public/modules/image/fixtures/image.component.fixture.html:
```diff
@@ -0,0 +1,7 @@
+<sky-image
+[showBorder]="showBorder"
```
The attributes need to be tabbed over once.
```
<sky-image
  [showBorder]="showBorder"
```

Blackbaud-SteveBrush commented on src/app/public/modules/image/fixtures/image.component.fixture.ts:
```diff
@@ -0,0 +1,17 @@
+import { Component } from '@angular/core';
```
Import statements need to follow code style guidelines.
See: https://docs.blackbaud.com/skyux-team-docs/style#imports

Blackbaud-SteveBrush commented on src/app/public/modules/image/image.component.scss:
```diff
@@ -0,0 +1,86 @@
+@import './node_modules/@skyux/theme/scss/mixins.scss';
```
I believe you can use the tilde here?
```
@import '~@skyux/theme/scss/mixins';
```

Blackbaud-SteveBrush commented on src/app/public/modules/image/image.component.ts:
```diff
@@ -0,0 +1,23 @@
+import { Component, Input } from '@angular/core';
```
Import statements need to follow code style guidelines.
See: https://docs.blackbaud.com/skyux-team-docs/style#imports

Blackbaud-SteveBrush commented on src/app/public/modules/image/image.component.ts:
```diff
@@ -0,0 +1,23 @@
+@Component({
+  selector: 'sky-image',
+  templateUrl: './image.component.html',
+  styleUrls: ['./image.component.scss']
```
Should define `changeDetection: ChangeDetectionStrategy.OnPush`.

Blackbaud-SteveBrush commented on src/app/public/modules/image/image.module.ts:
```diff
@@ -0,0 +1,22 @@
+import { NgModule } from '@angular/core';
```
Import statements need to follow code style guidelines.
See: https://docs.blackbaud.com/skyux-team-docs/style#imports

Blackbaud-SteveBrush commented on src/app/public/modules/image/index.ts:
```diff
@@ -0,0 +1,7 @@
+export {
```
Mind if we remove all `index.ts` files except for `./index.ts` and `./src/app/public/index.ts`? I need to update the style guide, but we're going to move away from using exported "barrel" files since they can easily cause circular import references.

Blackbaud-SteveBrush commented on src/app/public/modules/index.ts:
```diff
@@ -0,0 +1 @@
+export * from './image';
```
Remove this barrel file (see above).

Blackbaud-SteveBrush commented on src/app/public/modules/shared/image-resources.module.ts:
```diff
@@ -0,0 +1,20 @@
+    multi: true
+  }]
+})
+export class SkyImageResourcesModule { }
```
Since this will be used by all modules in this library, we should name it `SkyMediaResourcesModule`.

Blackbaud-SteveBrush commented on src/app/public/plugin-resources/image-resources-provider.ts:
```diff
@@ -0,0 +1,8 @@
+  SkyLibResourcesProvider
+} from '@skyux/i18n';
+
+export class SkyImageResourcesProvider implements SkyLibResourcesProvider {
```
Since this will be used by all modules in this library, we should name it `SkyMediaResourcesProvider`.

Blackbaud-SteveBrush commented on src/app/visual/image/image-visual.component.html:
```diff
@@ -0,0 +1,50 @@
+  <sky-image
+    imageSource="~/assets/demo-img.png"
+    imageAlt="image for visual test"
+    caption="default caption">
```
Closing `>` should be on its own line.
```
  caption="default caption"
>
</sky-image>
```
See: https://docs.blackbaud.com/skyux-team-docs/style#component-html-tags

Blackbaud-SteveBrush commented on src/app/visual/image/image-visual.component.ts:
```diff
@@ -0,0 +1,7 @@
+import { Component } from '@angular/core';
```
Import statements need to follow code style guidelines.
See: https://docs.blackbaud.com/skyux-team-docs/style#imports

Blackbaud-SteveBrush commented on src/assets/locales/resources_en_US.json:
```diff
@@ -0,0 +1,10 @@
+{
+  "sky_image_do_test": {
```
Should this be `"sky_image_do_text"`?

Blackbaud-SteveBrush replied:
Also, I know we're inconsistent, but the intended prefix for the strings is `skyux_`.
See: https://github.com/blackbaud/skyux-list-builder/blob/master/src/assets/locales/resources_en_US.json#L2

Blackbaud-SteveBrush commented on src/assets/locales/resources_en_US.json:
```diff
@@ -0,0 +1,10 @@
+{
+  "sky_image_do_test": {
+    "message": "Do",
+    "_description": "prefix text for success content"
```
Capitalize first letter in the `_description`.
`"Prefix text..."`
